Sources/BaggageContext/Context.swift

```diff
@@ -0,0 +1,287 @@
+/// Please refer to the "Reference Implementation" notes on each of the requirements to know how to implement this protocol correctly.
+///
+/// ### Relation to `BaggageProtocol`
+/// The `ContextProtocol` does NOT extend `BaggageProtocol`.
+///
+/// A custom context implementation however `MAY` extend both `ContextProtocol` and `BaggageProtocol`.
+/// Doing so enables a custom context to obtain the baggage protocol's capability of getting *and setting*
+/// properties on a baggage directly, like so: `myContext.traceID = "111-222"` without having to type out the baggage
+/// explicitly, as would normally be the case (`myContext.baggage.traceID = "111-222"`).
+///
+/// This difference in capabilities is purposefully designed as such, in order to allow context objects implement `ContextProtocol`
+/// without having to force them to allow _modifying_ the baggage. Some frameworks or libraries may choose to create immutable
+/// context objects, including the baggage they offer there. Users, when wanting to add more contextual information need then to
+/// perform the `asBaggageContext`, mutate the resulting context OR use the `with...` fluent-API functions to obtain a mutated context.
+///
+/// Please note that if a context does conform to `BaggageProtocol` as well, it will most-likely also offer all properties that
+/// users have provided via extensions on it - so be watchful of potential naming conflicts on variable names.
+public protocol ContextProtocol {
+    /// Get the `Baggage` container.
+    var baggage: Baggage { get }
+
+    /// The `Logger` associated with this context carrier.
+    ///
+    /// It automatically populates the loggers metadata based on the `BaggageContext` associated with this context object.
+    ///
+    /// ### Implementation note
+    ///
+    /// Libraries and/or frameworks which conform to this protocol with their "Framework Context" types,
+    /// SHOULD implement this logger by wrapping the "raw" logger associated with  `_logger.with(self.baggage)` function,
+    /// which efficiently handles the bridging of baggage to logging metadata values.
+    ///
+    /// ### Reference Implementation
+    ///
+    /// Writes to the `logger` metadata SHOULD NOT be reflected in the `baggage`,
+    /// however writes to the underlying `baggage` SHOULD be reflected in the `logger`.
+    ///
+    ///     struct MyFrameworkContext: ContextProtocol {
+    ///       var baggage: Baggage
+    ///       private let _logger: Logger
+    ///
+    ///       var logger: Logger {
+    ///         return self._logger.with(self.baggage)
+    ///       }
+    ///     }
+    var logger: Logger { get }
+
+    /// Get a baggage `Context` struct, regardless of the representation that this context protocol is implemented by.
+    var asBaggageContext: Context { get }
+}
+
+// ==== ----------------------------------------------------------------------------------------------------------------
+// MARK: `with...` functions
+
+extension Context {
```

Comment:
The `with...` functions are necessary for the API to not such when a context implementation is immutable.

Today's `Lambda.Context` is an example of such context -- it is on purpose an immutable _class_ (though it may change, we'll see... Discussed with @fabianfett a bit about this). In any case, if someone wants to have an immutable context -- they can do so and these withs help with passing things around then.

It is especially useful with `handler { context in client.call(context.withLogLevel(.trace) }` etc, small one-liner changes.

The same can be done with baggage values

Reply:
As many libraries will be passing around a `ContextProtocol` these `with..` functions will require a double copy. The first one, assuming you haven't been passed a `Context`, when you call `ContextProtocol.asBaggageContext` and the second one when you call the `with...` function. Not sure what the optimizer will do here, maybe one of the copies will disappear. Is this considered an issue? 

Would it not be better to implement these as extensions of `ContextProtocol` eg
```
extension ContextProtocol {
   public func withLogger(_ logger: Logger) -> Context {
        return Context(baggage: self.baggage, logger: logger)
    }
}

Reply:
> Would it not be better to implement these as extensions of ContextProtocol eg

This is troublesome in many ways:

1. unexpected return type `extension ContextProtocol { public func withLogger(_ logger: Logger) -> Context`

As a user of some framework when I see `(framework)context.withLogger(...)` I'd expect to get back `FrameworkContext`, because I'd want to use it like:

```swift
func frameworkFunction(context: FrameworkContext)

frameworkFunction(context: frameworkContext.withLogLevel(.debug))
```

right?

So this won't work if we return `-> Context` (or `ContextProtocol`).

--- 

2. So let's say we want to return `Self` which would then in the above case: `extension ContextProtocol { public func withLogger(_ logger: Logger) -> Context`... We can't implement this without forcing the ContextProtocol to require logger and baggage to be `{ get set }`.

The `set` requirement specifically has been voiced as annoying by some libraries (Lambda in this case); though it could be argued lambda's context may want to be a struct rather than an immutable class like it is today... Today it does not allow mutation, and implementing the `set` on either is painful and requires e.g. locking.

The goal of the Context protocol is to be _trivial_ to implement, to foster adoption and don't force our opinion too much on those libs. This imples we can't force that set requirement. And without `baggage/logger { get set }` we can't implement this:

```swift
extension ContextProtocol {
    public func withLogger(_ logger: Logger) -> Self {
        let copy = self
        copy._logger = logger // can't do this
        return copy
    }
}
```

// I guess it's even worse because we'd want to set the _logger and not logger to avoid double wraping... 

Summary of 1) and 2):

- we can't keep the Context protocol trivial to implement AND provide all those helpers,
- we can provide the helpers on the default context struct (which this PR does),
- framework authors MAY implement such `myContext.with...` if they want to but they don't have to,

Reply:
> As many libraries will be passing around a ContextProtocol these with.. functions will require a double copy. 
> 
> The first one, assuming you haven't been passed a Context, when you call `ContextProtocol.asBaggageContext` and the second one when you call the with... function. 

these libraries are absolutely free to implement `withLogLevel` functions DIRECTLY on their types though, or even -- that's the intent, to show the shape, implement it on the struct we offer and if libs want to they can offer this.

Because of 1) and 2) above, we can't automagically implement it for contexts.

Reply:
We could try by requiring some "copy initializer" perhaps... but that again makes the protocol non trivial to implement -- so do we really need to invest effort to make it more complex there? WYDT?

Reply:
I see your point about how `with...` functions should really return `Self`. Implementing it is a pain though. I've just been down that rabbit hole myself trying to implement a similar pattern for Soto AWS service structs. 

My current intention would be to not have a `SotoContext` but pass on what context is supplied to me, but I would like to be able to add to the baggage. This requires the conversion to `Context` which I guess I'll have to put up with.

Reply:
> I see your point about how with... functions should really return Self. Implementing it is a pain though. I've just been down that rabbit hole myself trying to implement a similar pattern for Soto AWS service structs.

Yeah, it's a world of pain we'd throw onto any library wanting to use context.

There's various ways to get it implemented, all are pretty painful (making a special init requirement, forcing every impl to implement those, etc). In reality I think we're best off just keeping it simple and proposing the shapes on the concrete type; others may or may not follow. It is "just" convenience API after all.

> My current intention would be to not have a SotoContext but pass on what context is supplied to me, 

That's the right choice 👍 In other words, APIs accepting context should always accept the `protocol Context`, if one starts demanding "my own" we're back in the various contexts hell. The entire protocol dance is really only meant for "servers" (a thing that parses request + request context and calls you with it, those can define their context since they create and call you with it). Client libraries should not really define their own types... 

So HTTPClient will also not define it's own -- this way indeed you'll be able to accept a context and pass it through "as is" to HTTPClient 👍 

> but I would like to be able to add to the baggage. This requires the conversion to DefaultContext which I guess I'll have to put up with.

Yeah... therefore a `var baggage { set get }` would really be useful... I have very mixed feelings on it... It's kind of the right thing but only if contexts would be guaranteed to be value types but they're often not (for better or worse) (e.g. vapor).

This is another reason why I had the `asBaggageContext.withBaggage { $0.something = "" }` prepared really; but since we can't nicely make it return `Self` I guess that's dead in the water as well.

Thus the entire "simple is better" vibe for this final api proposal. Going too smart just explodes in complexity very fast.

Open to ideas, but I think we'll be looping back to the simplest forms of APIs...

Reply:
> but I would like to be able to add to the baggage.

Could you let us know what types of values you want to add to the baggage btw?

I'm mostly curious about gathering all use cases people are thinking of

Reply:
I was going to add the metadata I was attaching to the Logger: AWS service and operation eg `S3` and `ListObjects`. I also had a request id attached to the logger, not sure that is needed anymore. But I would like to see how this in visible in the XRay, because we could also add a span called "S3.ListObjects".

What are your views on span names? When you discussed Logging you were keen to have static messages with the metadata holding the dynamic elements of the log message.

Reply:
@adam-fowler I think we should get the AWS naming pattern from go or node ;) Since tracing is cross language cutting, I don't think that it matters much what each language does, but more that you can a consistent view across languages.

Reply:
> What are your views on span names? When you discussed Logging you were keen to have static messages with the metadata holding the dynamic elements of the log message.

Span names can be descriptive and dynamic, after all that's what shows up in UIs usually on the "bar" so like

```
    [GET /thingy-----]
                      [MyHandler:processTheResult]
```

are all fine names and they're dynamic (paths, methods, etc)

Reply:
@fabianfett that makes sense.